sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -619,3 +619,34 @@ object DataSourceStrategy {
     (nonconvertiblePredicates ++ unhandledPredicates, pushedFilters, handledFilters)
   }
 }
+
+
+/**
+ * Defines default table statistics if table statistics are not available.
+ */
+class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+
+  private val conf = session.sessionState.conf
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    // For the data source table, we only recalculate the table statistics
```

Comment:
shall we only do it for partitioned tables? Seems non-partitioned table already uses the actual file size.

Reply:
How about we only do it when it will use `CatalogFileIndex`?
```scala
    case logical @ LogicalRelation(_, _, Some(table), _)
      if DDLUtils.isDatasourceTable(table) && table.stats.isEmpty &&
        conf.fallBackToHdfsForStatsEnabled && conf.manageFilesourcePartitions &&
        table.tracksPartitionsInCatalog && table.partitionColumnNames.nonEmpty =>
      val sizeInBytes = CommandUtils.getSizeInBytesFallBackToHdfs(session, table)
      val withStats = table.copy(stats = Some(CatalogStatistics(sizeInBytes = BigInt(sizeInBytes))))
      logical.copy(catalogTable = Some(withStats))
```

https://github.com/apache/spark/blob/5bb69945e4aaf519cd10a5c5083332f618039af0/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L382-L391

Reply:
Shouldn't we check the relation also has the stats? If then, should we fallback to hdfs instead of using that stats. Otherwise the same regression issue  in #22502 also will be here right?

Reply:
Do you mean `BaseRelation` in `LogicalRelation`?

Reply:
Yes. That was the issue in #22502 right? How is it solving in this case?

Reply:
We will not use the `sizeInBytes` of the  `BaseRelation` if table stats are available:
https://github.com/apache/spark/blob/a1c1dd3484a4dcd7c38fe256e69dbaaaf10d1a92/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L42-L46

Reply:
Yeah, then there will not be any behavior change between this PR and #22502. right?
Because, it it will check the table stats is empty or not at this line.
```
case logical @ LogicalRelation(_, _, Some(table), _)
      if DDLUtils.isDatasourceTable(table) && table.stats.isEmpty &&
```
In #22502, checking table stats exists here,
https://github.com/apache/spark/blob/f999e00e9fc03889106a305eeec6d05b6a2c1de1/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L43-L45

So, the fallback behavior is same in both the PRs, by not considering the relation stats and falling back to HDFS if the table size is empty?

Reply:
No, they are different. This pr get stats from `catalogTable.flatMap(_.stats.map(_.toPlanStats(output, conf.cboEnabled))).get`, but #22502 get stats from `Statistics(sizeInBytes = relation.sizeInBytes)`.

Reply:
Yes. agree. My point is, both the PRs, will always fallback to HDFS is the table statistics is empty and doesn't consider if the `catalogFileIndex` has stats or not.

Reply:
No. This pr only do it when:
https://github.com/apache/spark/blob/1fce50859b959cd4190f0da5dabf4addd187fb79/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L635-L637

Otherwise, we will get the default table size from `CatalogFileIndex`, which we constructed in `DataSource`:
https://github.com/apache/spark/blob/5bb69945e4aaf519cd10a5c5083332f618039af0/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L386-L390

Reply:
Yes. Both the PR checks, if using  `CatalogFileIndex` and table stats is empty, then fallback to HDFS to get the statistics. And both the PRs never return statistics from `CatalogFileIndex` irrespective of the stats is empty or not, if enable fallback. That is why I said the behavior is same.

 The pr is looks good. only concern is, if the catalogFileIndex already has statistics, then in the rules, we shouldn't fallback to HDFS, because that was the issue with #22502. Thanks.